### PR TITLE
[opentitantool] add a command to export a public key to header file

### PR DESCRIPTION
**Pull request by pamaury**
The "show" command can be used to display a key but the output formats supported are only json, json5, etc. It does not make sense to add an output format just for the ROM since it is so specific to this particular command. Instead I created another command that output the header that can be used in the ROM/ROM_EXT.

### CODE REVIEW

File: sw/host/opentitantool/src/command/rsa.rs [hunk 1]
```diff
@@ -16,6 +19,17 @@ use opentitanlib::crypto::rsa::{
 use opentitanlib::crypto::sha256::Sha256Digest;
 use opentitanlib::util::parse_int::ParseInt;
 
+/// Load a public key, or a private key and extract a public key from it
```
Comment (jrvanwhy) [inline sw/host/opentitantool/src/command/rsa.rs:22]: This comment could be a bit more clear:

```rust
/// Given the path to a public key, returns the public key. Given
/// the path to a private key, extracts the public key from the private
/// key and returns the public key.
```

File: sw/host/opentitantool/src/command/rsa.rs [hunk 2]
```diff
@@ -16,6 +19,17 @@ use opentitanlib::crypto::rsa::{
...
+fn load_pub_or_priv_key(path: &PathBuf) -> Result<RsaPublicKey> {
+    match RsaPublicKey::from_pkcs1_der_file(path) {
+        Ok(key) => Ok(key),
+        Err(_) => match RsaPrivateKey::from_pkcs8_der_file(path) {
+            Ok(key) => Ok(RsaPublicKey::from_private_key(&key)),
+            Err(e) => Err(e),
+        },
+    }
```
Comment (jrvanwhy) [inline sw/host/opentitantool/src/command/rsa.rs:30]: You can replace the inner `match` with `?`:

```rust
fn load_pub_or_priv_key(path: &PathBuf) -> Result<RsaPublicKey> {
    match RsaPublicKey::from_pkcs1_der_file(path) {
        Ok(key) => Ok(key),
        Err(_) => Ok(RsaPublicKey::from_private_key(RsaPrivateKey::from_pkcs8_der_file(path)?)),
    }
}
```

You can then de-duplicate the `Ok()` invocations:

```rust
fn load_pub_or_priv_key(path: &PathBuf) -> Result<RsaPublicKey> {
    Ok(match RsaPublicKey::from_pkcs1_der_file(path) {
        Ok(key) => key,
        Err(_) => RsaPublicKey::from_private_key(RsaPrivateKey::from_pkcs8_der_file(path)?),
    })
}
```

Alternatively, you can return early for a public key (I personally find this a bit nicer but it's subjective):

```rust
fn load_pub_or_priv_key(path: &PathBuf) -> Result<RsaPublicKey> {
    if let Ok(key) = RsaPublicKey::from_pkcs1_der_file(path) {
        return Ok(key);
    }
    Ok(RsaPublicKey::from_private_key(RsaPrivateKey::from_pkcs8_der_file(path)?))
}
```

File: sw/host/opentitantool/src/command/rsa.rs [hunk 3]
```diff
@@ -98,10 +106,126 @@ impl CommandDispatch for RsaKeyGenerateCommand {
...
+    #[structopt(
+        name = "DER_FILE",
+        help = "RSA public or private key file in DER format"
+    )]
+    der_file: PathBuf,
+    #[structopt(name = "OUTPUT_FILE", help = "output header file to generate")]
+    output_file: Option<PathBuf>,
+}
+
+/// Write the content of a big integer to a big as an array of 32-bit words.
```
Comment (jrvanwhy) [inline sw/host/opentitantool/src/command/rsa.rs:122]: It's unclear to me what the second "big" means on this line, please clarify (emphasized below):

Write the content of a big integer to a **big** as an array of 32-bit words.

File: sw/host/opentitantool/src/command/rsa.rs [hunk 4]
```diff
@@ -98,10 +106,126 @@ impl CommandDispatch for RsaKeyGenerateCommand {
...
+        // print prefix and newline on multiples of nr_per_line
+        if idx % nr_per_line == 0 {
+            if idx != 0 {
+                write!(out, "{}\n", suffix)?;
+            }
+            write!(out, "{}", prefix)?;
+        }
+        // print one 32-bit word
+        let val =
+            num[0] as u32 | (num[1] as u32) << 8 | (num[2] as u32) << 16 | (num[3] as u32) << 24;
```
Comment (jrvanwhy) [inline sw/host/opentitantool/src/command/rsa.rs:151]: Would `u32::from_le_bytes` or `u32::from_be_bytes` be helpful here?

File: sw/host/opentitantool/src/command/rsa.rs [hunk 5]
```diff
@@ -98,10 +106,126 @@ impl CommandDispatch for RsaKeyGenerateCommand {
...
+        write!(out, "{:#010x}, ", val)?;
+    }
+    // extra return to the line
+    write!(out, " {}\n", suffix)?;
+    Ok(())
+}
```
Comment (jrvanwhy) [inline sw/host/opentitantool/src/command/rsa.rs:157]: `write_bigint_as_u32` could benefit from a unit test. The test would also be helpful as an example (i.e. it would show an example input and an example output).

File: sw/host/opentitantool/src/command/rsa.rs [hunk 6]
```diff
@@ -98,10 +106,126 @@ impl CommandDispatch for RsaKeyGenerateCommand {
...
+                // if no output file is provided, derive the name of the file from the
+                // name of the key
+                let mut out_path = self.der_file.clone();
+                out_path.set_extension("h");
+                out_path
+            }
+        };
+        // as a security precaution, we never allow the output file to overwrite the key file
+        if self.der_file == output_path {
+            bail!("the output file is the same as the key file, this would overwrite the key, not allowing this")
```
Comment (jrvanwhy) [inline sw/host/opentitantool/src/command/rsa.rs:179]: This check doesn't seem super robust, e.g. I don't think it follows symlinks. Do we need to make this more robust?

File: sw/host/opentitantool/src/command/rsa.rs [hunk 7]
```diff
@@ -98,10 +106,126 @@ impl CommandDispatch for RsaKeyGenerateCommand {
...
+        let mut file = File::create(&output_path)?;
+        // construct a key name from the key file name
+        let keyname = self
+            .der_file
+            .file_name()
+            .expect("this should be a valid file name since we opened the file")
+            .to_string_lossy();
+        let re = Regex::new(r#"(.pub)?.der$"#).expect("this should be a valid regex");
+        let keyname = re.replace_all(&keyname, "");
+        let re = Regex::new(r#"[^a-zA-Z0-9]"#).expect("this should be a valid regex");
```
Comment (jrvanwhy) [inline sw/host/opentitantool/src/command/rsa.rs:192]: `"this should be a valid regex"` isn't a very informative error message. Since we assume these should never fail, they should probably just be `unwrap`s.

File: sw/host/opentitantool/src/command/rsa.rs (hunk 2 above)
Comment (pamaury) [reply sw/host/opentitantool/src/command/rsa.rs:30]: Thanks, I went for the early return, I also find it nicer.

File: sw/host/opentitantool/src/command/rsa.rs (hunk 4 above)
Comment (pamaury) [reply sw/host/opentitantool/src/command/rsa.rs:151]: Yes. I am not entirely happy with the solution, I went for `u32::from_le_bytes(num.try_into().unwrap());`.
The alternative would be to `u32::from_le_bytes([num[0], num[1], num[2], num[3]]);` but at least the first one makes it clear why we are chunking the bytes together.

File: sw/host/opentitantool/src/command/rsa.rs (hunk 6 above)
Comment (pamaury) [reply sw/host/opentitantool/src/command/rsa.rs:179]: The idea was to prevent "silly" mistakes like typing the same path twice, forgetting to replace the extension. Checking for file equality is a bit tricky, if we want to do it robustly then it probably best to use the `same-file` crate.

File: sw/host/opentitantool/src/command/rsa.rs (hunk 3 above)
Comment (pamaury) [reply sw/host/opentitantool/src/command/rsa.rs:122]: That was a typo, indeed it made the text super unclear.

File: sw/host/opentitantool/src/command/rsa.rs (hunk 5 above)
Comment (pamaury) [reply sw/host/opentitantool/src/command/rsa.rs:157]: Yes, I will add one to the next update of the PR.

Comment (pamaury) [reply sw/host/opentitantool/src/command/rsa.rs:155]: I have added a unittest here (I don't think it is necessary to move this function to the opentitanlib?).
Hopefuly the test clarifies what the function is doing.

File: sw/host/opentitantool/src/command/rsa.rs (hunk 6 above)
Comment (jrvanwhy) [reply sw/host/opentitantool/src/command/rsa.rs:195]: In that case, maybe rephrase "as a security precaution" to something like "To catch mistakes" or something like that. "security" conveys a higher level of importance than this check has.

**Review by jrvanwhy — APPROVED** [review]
Looks good from a Rust perspective.

**Comment by jrvanwhy** [comment]
I forgot to mention: looks like you just need to re-run `rustfmt` to make CI happy.

**Comment by pamaury** [comment]
I have updated the PR. I also added a comment about bazel stamping since it might not be obvious (it definitely was not to me when I first looked into running the unittest).

**Review by moidx — APPROVED** [review]

**Review by engdoreis — APPROVED** [review]
